Add new ChatGPT integration article (#179)
ScottLogic/blog: colive/_posts/2024-04-08-how-to-make-chatgpt-do-anything-you-want.md

---
author: colive
title: How to make ChatGPT do anything you want
layout: default_post

> summary: Using an example project this article explore the possibilities of integrating a Large Language Model with custom logic to improve it's utility and power.

Reviewer: ```suggestion
summary: Using an example project this article explores the possibilities of integrating a Large Language Model with custom logic to improve its utility and power.
```

category: 
    - Artificial Intelligence
tags: 
    - ChatGPT
    - Large Language Models
--- 

## Introduction

> OpenAI released an early demo of ChatGPT on the 30th of November 2022. Since then, it has propelled the general public’s awareness of the power of Large Language Models (LLM). In other words, ChatGPT was the first viral ChatBot. These Generative Pre-Trained Transformers take a prompt (a block of written text with no other context) and will attempt at producing a coherent written response. GPT models are trained on conversation data to learn the most appropriate response to whatever new prompt they are served, but it should be noted that the training data they have been built from is all they know. They are often compared to parrots due to this behavioural trait, echoing previously seen conversations to construct sensible responses.

Reviewer 2: I dont believe chatGPT was the first viral chatbot as they have been around a long time. Maybe should be changed to first viral AI chatbot or something similar

Author: Added "AI-powered"

> By itself a tool that can do this is very powerful. It's currently commonplace on social media to see posts around using models like GPT-3.5 (the current ChatGPT model used "under the hood") to generate templates for emails, social media posts, poems, birthday party ideas and all sorts of other supposedly creative ideas. However, the current limitation of most of these models is integration into other systems and knowledge bases. A pure LLM won't know the result of the football game last week or what Elon Musk last posted. Achieving this is something the large tech companies are actively working on, but it can also be achieved ourselves with very little code!

Reviewer: ```suggestion
By itself a tool that can do this is very powerful. It's currently commonplace on social media to see posts around using models like GPT-3.5 (the current ChatGPT model used "under the hood") to generate templates for emails, social media posts, poems, birthday party ideas and all sorts of other supposedly creative ideas. However, the current limitation of most of these models is integration into other systems and knowledge bases. A pure LLM won't know the result of the football game last week or what Elon Musk last posted. Achieving this is something the large tech companies are actively working on, but it can also be achieved by ourselves with very little code!
```

In this article I will be using my personal project [PunditBot](https://github.com/Ch0live/pundit-bot) as an example of how this can work, but the concept can be applied to anything. To keep things as simple as possible, let’s talk about achieving the goal we mentioned before - finding out the result of the football game last week.

## Using OpenAI's API

Currently if we ask ChatGPT about this it admits it doesn't have access to current data around football games - as we've established the model doesn't actively search the web for data to integrate into its response.

> ![Asking ChatGPT about the game]({{ site.github.url }}/colive/assets/askingChatGPT.png "Asking ChatGPT about the game")

Author: It _should_ do, I need to verify it first by generating the site on my own machine

> ![Asking ChatGPT about the game]({{ site.github.url }}/colive/assets/askingChatGPT.png "Asking ChatGPT about the game")

Author: In my latest commit the post renders everything correctly 👍

> We need to build new logic around the call to OpenAI's GPT model. Fortunately, OpenAI also offers requests straight to their model through an API, billed by the number of tokens (~3/4 of a word) generated in each response to your account. Now we can write code (here is a short python script as an example stolen directly from [OpenAI's Quickstart tutorial](https://platform.openai.com/docs/quickstart?context=python)) to call the API with our prompt.

Reviewer 3: is it billed just on the response, I thought I read somewhere that is was also the tokens we pass into the model?

Author: [It is both](https://help.openai.com/en/articles/6614209-how-do-i-check-my-token-usage)! Amended

```python
ChatCompletions
from openai import OpenAI
client = OpenAI()

completion = client.chat.completions.create(
  model="gpt-3.5-turbo",
  messages=[
    {"role": "system", "content": "You are PunditBot, aimed at reporting on football games"},
    {"role": "user", "content": "Who won the man city arsenal game last weekend?"}
  ]
)

print(completion.choices[0].message)
```

When using the `/v1/chat/completions` endpoint you can specify both a system prompt (response context for the LLM) and a user prompt (the actual prompt given, often called utterance).

## Building from others' work

> As with all programming endeavours lets aim not to re-invent the wheel and use an example project to use as a template for our own. In PunditBot I used [the docker genai-stack project](https://github.com/docker/genai-stack). Note that there are many other repositories that can be great for starting a project like this from, this is just the one I chose.

Reviewer: ```suggestion
As with all programming endeavours let's aim not to re-invent the wheel and use an example project to use as a template for our own. In PunditBot I used [the docker genai-stack project](https://github.com/docker/genai-stack). Note that there are many other repositories that can be great for starting a project like this from, this is just the one I chose.
```

This repository contains other interesting tools that utilize an LLM model, but most importantly it has a Vite front end application connected to a FastAPI backend, with configuration we can plug our new OpenAI API key into. 

So now we have our own version of ChatGPT - a chat bot interface that uses an LLM for generated answers. A simplified diagram of the app architecture would look like this.

![GenAI template architecture]({{ site.github.url }}/colive/assets/basic.png "GenAI template architecture")

## Creating a bespoke chatbot

Our app is still unable to report on the game last week. However, our app now has a backend layer we can tailor to whatever needs we wish. One way to implement this logic could be the following

1. Take the question from the user and extract the football teams mentioned
2. Look up the last game played between the teams mentioned
3. Return a coherent response to the user containing that game data we extracted in step 2.

> To complete these steps, we are going to build an extractor function to identify the teams mentioned, a lookup function to find the data for the teams last game and a summarizer function to bundle up the data into a nicely written response. Exploring step 2 isn't of relevance to this article as it does not involve an LLM. We could use a local database and query it in SQL, or a tool that searches the web or requests data from football API for the game results. Either way this functionality is not novel or new; lets instead explore steps 1 and 3.

Reviewer 3: Should the `teams last game` be `teams' last game` ?

> To complete these steps, we are going to build an extractor function to identify the teams mentioned, a lookup function to find the data for the teams last game and a summarizer function to bundle up the data into a nicely written response. Exploring step 2 isn't of relevance to this article as it does not involve an LLM. We could use a local database and query it in SQL, or a tool that searches the web or requests data from football API for the game results. Either way this functionality is not novel or new; lets instead explore steps 1 and 3.

Reviewer: ```suggestion
To complete these steps, we are going to build an extractor function to identify the teams mentioned, a lookup function to find the data for the teams' last game and a summariser function to bundle up the data into a nicely written response. Exploring step 2 isn't of relevance to this article as it does not involve an LLM. We could use a local database and query it in SQL, or a tool that searches the web or requests data from football API for the game results. Either way this functionality is not novel or new; let's instead explore steps 1 and 3.
```

```
You are an expert extraction algorithm tasked at extracting what football match (or matches) the user wants to know about. You must return the 2 football clubs that the user has asked about from the question. Return the 2 clubs in a comma separated list.
	
Football clubs must be referred to using one of the following names

	- Arsenal FC
	- Brighton & Hove Albion FC
	- Chelsea FC
	- Crystal Palace FC
	- Everton FC
	- Southampton FC
	- Watford FC
	- West Bromwich Albion FC
	- Manchester United FC
	- Newcastle United FC
	- AFC Bournemouth
	- Burnley FC
	- Leicester City FC
	- Liverpool FC
	- Stoke City FC
	- Swansea City AFC
	- Huddersfield Town AFC
	- Tottenham Hotspur FC
	- Manchester City FC
	- West Ham United FC

For example, for the question

"What happened the last time arsenal played spurs?"

You should return 

"Arsenal FC, Tottenham Hotspur FC"

If you do not know the value of an attribute asked to extract, return null for the attribute's value.

The question: {question}
```

> This prompt is engineered to ask the LLM for a specific outcome - find from the question provided two football teams. The specific football clubs that should be returned are listed (this data is from Premier League teams during the 2017/18 season in case you are wondering why Huddersfield Town are there). There is also an example of One-Shot Learning - where an example model answer has been provided. Finally, it is mentioned that if the answer cannot be found the LLM should return null and not hallucinate a response.

Reviewer 3: Don't know whether this is too nit picky, but you use hallucinate here for the first time and then describe what it means when talking about misinformation later on. Wonder whether it should be the other way around

Author: Yes, best to describe it on first mention. I have moved that explanation to here

This works surprisingly well! Beyond clear questions specifying the names of the teams, GPT-3.5 was able to identify the two teams involved in the Manchester Derby and could even understand "the blues" as Chelsea FC.

## Creating a summarizer method

After step 1 and step 2 we need to return to the user the game data found in a Chat Bot style response. This again is remarkably simple using an LLM, see the example prompt below.

```
You have been asked to find the results of matches between the following Premier League football teams

here is data about the games in a python dictionary format

{matchData}

Summarise a response to the question using the data listed above.
Only use data listed above. 
Do not infer any data outside of what is listed above.
```

> Like the extractor prompt, the content within the curly brackets will contain the actual match data. In my example I injected the string representation of a python dictionary containing key value pairs of game data (half-time score, home team, date of the match, etc.). We heavily emphasize in the end of the prompt to only use mentioned data to avoid the potential of misinformation (an LLM returning misinformation confidently as fact is often described as the LLM hallucinating).

Reviewer: ```suggestion
Like the extractor prompt, the content within the curly brackets will contain the actual match data. In my example I injected the string representation of a python dictionary containing key value pairs of game data (half-time score, home team, date of the match, etc.). We heavily emphasise in the end of the prompt to only use mentioned data to avoid the potential of misinformation (an LLM returning misinformation confidently as fact is often described as the LLM hallucinating).
```

![Asking PunditBot about the game]({{ site.github.url }}/colive/assets/askingPunditBot.png "Asking PunditBot about the game")

Now there are obvious flaws with the current version I am showcasing above.

- As I am using a static database of games from just 1 season PunditBot is only able to retrieve data from those games to find out the "last" game that occurred
	- This can be solved by adding a better way of getting the data
- The UI includes unnecessary information around if "RAG mode" is enabled or not or what model is being used
	- These are just parts of the UI that came with [the docker genai-stack project](https://github.com/docker/genai-stack) I based the project from and can be cleaned up
- The response is poorly formatted and hard to read
	- Some tweaking of the summarizer prompt could be used to enforce new lines between the 2 game reports, plus other visual adjustments if necessary.

> Ultimately this article is concerned with the application of this technology. What has been proven here is it is possible to take an LLM model like GPT-3.5 and add new functionality around it to make it more useful for specific applications. There are frameworks (such as [LangChain](https://www.langchain.com/)) that are entirely based on this idea. The idea could be taken a lot further by plugging in more tools - for example we could return a video highlight of the winning goal to both games and render that beneath the game reports.

Reviewer 3: Used `this` twice in a sentence for two meanings, also I think a comma is needed after `Ultimately`. What about something like: `Ultimately, this article is concerned with the application of utilising an LLM with external tools` ?
